extension/src/openvic-extension/singletons/SoundSingleton.cpp

```diff
@@ -0,0 +1,430 @@
+using OpenVic::Utilities::std_to_godot_string;
+
+using namespace godot;
+using namespace OpenVic;
+using namespace OpenVic::NodeTools;
+
+//ov_bind_method is used to make a method visible to godot
+void SoundSingleton::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("load_music"), &SoundSingleton::load_music);
+	ClassDB::bind_method(D_METHOD("get_song","song_name"), &SoundSingleton::get_song);
+	ClassDB::bind_method(D_METHOD("get_song_list"), &SoundSingleton::get_song_list);
+	
+	ADD_PROPERTY(PropertyInfo(
+		Variant::ARRAY,
+		 "song_list", PROPERTY_HINT_ARRAY_TYPE,
+		 "AudioStreamMP3"),
+	"", "get_song_list");
+
+
+	ClassDB::bind_method(D_METHOD("load_sounds"), &SoundSingleton::load_sounds);
+	ClassDB::bind_method(D_METHOD("get_sound_stream","sound_name"), &SoundSingleton::get_sound_stream);
+	ClassDB::bind_method(D_METHOD("get_sound_base_volume","sound_name"), &SoundSingleton::get_sound_base_volume);
+	ClassDB::bind_method(D_METHOD("get_sound_list"), &SoundSingleton::get_sound_list);
+	
+	ADD_PROPERTY(PropertyInfo(
+		Variant::ARRAY,
+		"sound_list", 
+		PROPERTY_HINT_ARRAY_TYPE,
+		"AudioStreamWAV"),
+	"", "get_sound_list");
+	
+	ClassDB::bind_method(D_METHOD("load_title_theme"), &SoundSingleton::load_title_theme);
+	ClassDB::bind_method(D_METHOD("get_title_theme"), &SoundSingleton::get_title_theme);
+	
+	ADD_PROPERTY(PropertyInfo(
+		Variant::STRING,
+		"title_theme"),
+	"", "get_title_theme"); 
+
+}
+
+SoundSingleton* SoundSingleton::get_singleton() {
+	return _singleton;
+}
+
+SoundSingleton::SoundSingleton() {
+	ERR_FAIL_COND(_singleton != nullptr);
+	_singleton = this;
+}
+
+SoundSingleton::~SoundSingleton() {
+	ERR_FAIL_COND(_singleton != this);
+	_singleton = nullptr;
+}
+
+
+//Load a sound from the path 
+Ref<AudioStreamMP3> SoundSingleton::_load_godot_mp3(String const& path) const {
+	const Ref<FileAccess> file = FileAccess::open(path, FileAccess::ModeFlags::READ);
+	
+	Error err = FileAccess::get_open_error();
+	ERR_FAIL_COND_V_MSG(
+		err != OK || file.is_null(), nullptr,
+		vformat("Failed to open mp3 file %s", path) //named %s, path, 
+	);
+
+	const PackedByteArray data = file->get_buffer(file->get_length());
+
+	Ref<AudioStreamMP3> sound = Ref<AudioStreamMP3>();
+	sound.instantiate();
+	sound->set_data(data);
+
+	return sound;
+}
+
+//only used for music
+godot::String SoundSingleton::godot_path_to_file_name(String const& path) const {
+	String name = path.replace("\\","/");
+	return name.get_slice("music/",1).get_slice(".",0); 
+}
+
+//only used for sounds (.wav sfx)
+godot::String SoundSingleton::godot_path_to_file_name_extension(String const& path) const {
+	String name = path.replace("\\","/");
+	return name.get_slice("sound/",1); //get file name with extension
+}
+
+//Load a sound from the cache, or directly if its not in the cache
+//take in a path, extract just the file name for the cache (and defines)
+Ref<AudioStreamMP3> SoundSingleton::get_song(String const& path){
+	String name = godot_path_to_file_name(path);
```

Comment:
This is to do with the SongChance and SoundEffect defines. I wanted to load the map in a way that I could access an audio file based on how it was written in the defines. Right now for example, there may be a song `song = {name = "warFolder/track1.mp3" ...}` godot_path_to_file_name lets us access this audiostream using `name` as the key in the map, rather than generating a full file path `who/knows/what/victoria II/music/warFolder/...` to access the audio stream.

Reply:
Something that could be done to simplify things a bit would be to settle on using the file extension on both the music and sound folders, and combining the 2 rather similar functions. They were originally more different, but I ended up simplifying them to this.

Reply:
New commit combines the functions into "to_define_file_name", with the caveat that it runs a std_to_godot_string every time is runs. Added a comment explaining why it exists.